httpx/concurrency/trio.py

```diff
@@ -41,23 +39,19 @@ def get_http_version(self) -> str:
         ident = self.stream.selected_alpn_protocol()
         return "HTTP/2" if ident == "h2" else "HTTP/1.1"
 
-    async def read(
-        self, n: int, timeout: TimeoutConfig = None, flag: TimeoutFlag = None
-    ) -> bytes:
-        if timeout is None:
-            timeout = self.timeout
+    async def read_or_timeout(self, n: int, timeout: typing.Optional[float]) -> bytes:
+        with trio.move_on_after(_or_inf(timeout)):
+            return await self.stream.receive_some(max_bytes=n)
+        raise ReadTimeout()
 
-        while True:
-            # Check our flag at the first possible moment, and use a fine
-            # grained retry loop if we're not yet in read-timeout mode.
-            should_raise = flag is None or flag.raise_on_read_timeout
-            read_timeout = _or_inf(timeout.read_timeout if should_raise else 0.01)
-
-            with trio.move_on_after(read_timeout):
-                return await self.stream.receive_some(max_bytes=n)
-
-            if should_raise:
-                raise ReadTimeout() from None
+    async def write_or_timeout(
+        self, data: bytes, timeout: typing.Optional[float]
+    ) -> None:
+        with trio.move_on_after(_or_inf(timeout)):
+            async with self.write_lock:
+                await self.stream.send_all(data)
```

Comment:
Similarly to the asyncio write method, I don't understand how looping on calls to `send_all` is safe?

From the Trio docs for `send_all()`:
> Most low-level operations in Trio provide a guarantee: if they raise trio.Cancelled, this means that they had no effect, so the system remains in a known state. This is not true for send_all(). If this operation raises trio.Cancelled (or any other exception for that matter), then it may have sent some, all, or none of the requested data, and there is no way to know which.

How can we loop on this while guaranteeing that we don't send duplicate data?

Reply:
So, while asyncio has a write/drain pair that *could* allow us to safely retry waiting for the data to be sent, it seems trio has no such semantics, right? Should we just *never* retry on write?

Reply:
>  it seems trio has no such semantics, right?

_Technically_ what I think you _could_ do (although it might be horrible/complicated) is to shield the future returned by the write operation so that the timeout doesn't cancel it and then either cancel the future if you're in write timeout mode or await it if not. I think that would essentially mean _not_ doing this PR and keeping the retry logic in the specific backends.

> Should we just never retry on write?

I don't understand well enough the original intention of this read/write timeout logic to say.

Reply:
As it stands, even in the current state the trio backend is affected by this potential catch because we're looping on `send_all()`.

I'm going to cc @tomchristie because I know he's been wanting to review the timeout logic for a while.